bot/exts/valentines/markov_poem_generator.py

```diff
@@ -0,0 +1,513 @@
+            if word not in cache.keys():
+                cache[word] = await func(instance, word)
+
+        return cache[word]
+
+    return wrapper
+
+
+def command_wrapper(func: Callable) -> Callable:
+    """The decorator adds a timeout and shows the bot's `typing...` text."""
+    @functools.wraps(func)
+    async def wrapper(
+        *args,
+        instance: Callable = None,
+        timeout: int = None,
+        ctx: commands.Context = None,
+        scheme: str = None,
+        unit_count: Dict[str, int] = None,
+        time_start: datetime.time = None,
+        is_first_error: bool = None
+    ) -> Callable:
+        try:
+            async with async_timeout.timeout(timeout), ctx.typing():
+                return await func(instance, ctx, scheme, unit_count,
+                                  time_start, is_first_error)
+        except TimeoutError:
+            await ctx.send(f"Sorry {ctx.author.mention}, but your poem"
+                           " timed out, please try again!")
+
+    return wrapper
+
+
+class MarkovPoemGenerator(commands.Cog):
+    """
+    A cog that provides a poem by taking the markov of a corpus.
+
+    By processing the corpus text through a markov chain, a series of lines
+    can be iterated through, whilst corresponding to the given rhyme scheme.
+    """
+
+    POEM_TIMEOUT = 30  # In seconds
+    near_rhyme_min_score = 2000
+    rhyming_line_finder_limiter = 20000
+    max_char_range = (50, 120)  # For the sentence generator
+
+    SOURCES: List[str] = [
+        "shakespeare_corpus.txt"
+    ]
+
+    templates: Dict[str, str] = {
+        "shakespearean-sonnet": "abab/cdcd/efef/gg",
+        "spenserian-sonnet": "abab/bcbc/cdcd/ee",
+        "petrarch-sonnet": "abbaabba/cdecde",
+        "ballade": "ababbcbc",
+        "terza-rima": "aba/bcb/cdc/ded/ee",
+        "villanelle": "aba/aba/aba/aba/aba/abaa",
+        "limerick": "aabba"
+    }
+
+    rhyme_websites: List[Tuple[bool, str]] = [
+        # (is exact rhyme, website link)
+        (True, "https://api.datamuse.com/words?rel_rhy="),
+        (False, "https://api.datamuse.com/words?rel_nry=")
+    ]
+
+    def __init__(self, bot: commands.Bot):
+        """Initializes the full corpus text and the markov model."""
+        self.bot = bot
+
+        # Load the full text corpus
+        full_corpus = []
+        for source_file in self.SOURCES:
+            with Path(f"bot/resources/valentines/{source_file}").open() as f:
+                curr_corpus = f.read().splitlines()
+                full_corpus.extend(curr_corpus)
+
+        # Create the markov model
+        self.model = markovify.Text(full_corpus, state_size=1)
+
+        logging.info("Full text corpus and markov model successfully loaded.")
+
+    @staticmethod
+    def _get_unit_count(scheme: str) -> Dict[str, int]:
+        """Checks how many times a unit occurs in a rhyme scheme."""
+        return {char: scheme.count(char) for char in set(scheme)}
```

Comment:
Wouldn't this be a nice place to use [`collections.Counter`](https://docs.python.org/3/library/collections.html#collections.Counter)?

Reply:
@anand2312 I compared the two functions, and I think the current one is faster.